File: fsharp/FsSearch/SearchResult.fs
namespace FsSearch

open System
open System.Collections.Generic
open System.IO
open System.Text
open System.Text.RegularExpressions

module SearchResult =
    type t = {
        SearchPattern : Regex;
        File : SearchFile.t;
        LineNum : int;
        MatchStartIndex : int;
        MatchEndIndex : int;
        Line : string;
        LinesBefore : string list;
        LinesAfter : string list;
    }

    let Create (pattern : Regex) (lineNum : int) (startIndex : int) (endIndex : int) (line : string) (linesBefore : string list) (linesAfter : string list) : t =
        {
            SearchPattern = pattern;
            File = { Containers=[]; File=null; FileType=FileType.Unknown; };
            LineNum = lineNum;
            MatchStartIndex = startIndex;
            MatchEndIndex = endIndex;
            Line = line;
            LinesBefore = linesBefore;
            LinesAfter = linesAfter;
        }

    let SingleLineToString (sr : t) : string =
        let matchString =
            if sr.LineNum = 0 then
                sprintf " matches at [%d:%d]" sr.MatchStartIndex sr.MatchEndIndex
            else
                sprintf ": %d: [%d:%d]: %s" sr.LineNum sr.MatchStartIndex sr.MatchEndIndex (sr.Line.Trim())
        (SearchFile.ToString sr.File) + matchString

    let MultLineToString (sr : t) : string =
        let hdr = 
            String.concat "\n" [
                sprintf "%s" (new String('=', 80));
                sprintf "%s: %d: [%d:%d]" (SearchFile.ToString sr.File) sr.LineNum sr.MatchStartIndex sr.MatchEndIndex;
                sprintf "%s" (new String('-', 80));
            ] + "\n"
        let maxLineNum = sr.LineNum + (List.length sr.LinesAfter)
        let maxLineNumLength = maxLineNum.ToString().Length
        let paddedLineNum (lineNum : int) : string =
            let lineNumString = lineNum.ToString()
            (new String(' ', (maxLineNumLength - lineNumString.Length))) + lineNumString
        let rec recLines (lines : string list) (lineNum : int) (linesString : string) : string =
            match lines with
            | [] -> linesString
            | l :: ls ->
                recLines ls (lineNum + 1) (linesString + (sprintf "  %s | %s\n" (paddedLineNum lineNum) l))
        let linesBeforeString = recLines sr.LinesBefore (sr.LineNum - (List.length sr.LinesBefore)) ""
        let linesAfterString = recLines sr.LinesAfter (sr.LineNum + 1) ""
        String.concat "" [
            hdr;
            linesBeforeString;
            sprintf "> %s | %s\n" (paddedLineNum sr.LineNum) sr.Line;
            linesAfterString;
        ]

    let ToString (sr : t) : string =
        if (not (List.isEmpty sr.LinesBefore)) || (not (List.isEmpty sr.LinesAfter)) then
            MultLineToString sr
        else
            SingleLineToString sr
;;

File: ocaml/mlsearch/src/searchresult.ml
open Core.Std

type t = {
  search_pattern : string;
  file : Searchfile.t;
  line_num : int;
  match_start_index : int;
  match_end_index : int;
  line : string;
  lines_before : string list;
  lines_after : string list;
}

let create (pattern : string) (line_num : int) (start_index : int) (end_index : int) (line : string) (lines_before : string list) (lines_after : string list) : t =
  {
    search_pattern = pattern;
    file = { containers=[]; path=""; file_name=""; file_type=Filetypes.Unknown; };
    line_num = line_num;
    match_start_index = start_index;
    match_end_index = end_index;
    line = line;
    lines_before = lines_before;
    lines_after = lines_after;
  };;

let single_line_to_string (sr : t) = 
  let match_string =
    if sr.line_num = 0 then
      sprintf " matches at [%d:%d]" sr.match_start_index sr.match_end_index
    else
      sprintf ": %d: [%d:%d]: %s" sr.line_num sr.match_start_index sr.match_end_index (String.strip sr.line) in
  (Searchfile.to_string sr.file) ^ match_string;;

let multi_line_to_string (sr : t) = 
  let hdr = 
    String.concat [
      sprintf "%s\n" (String.make 80 '=');
      sprintf "%s: %d: [%d:%d]\n" (Searchfile.to_string sr.file) sr.line_num sr.match_start_index sr.match_end_index;
      sprintf "%s\n" (String.make 80 '-');
      ] in
  let max_line_num = sr.line_num + (List.length sr.lines_after) in
  let max_line_num_length = String.length (sprintf "%d" max_line_num) in
  let padded_line_num line_num = 
    let line_num_string = sprintf "%d" line_num in
    (String.make (max_line_num_length - (String.length line_num_string)) ' ') ^ line_num_string in
  let rec rec_lines lines line_num line_string = 
    match lines with
    | [] -> line_string
    | l :: ls ->
      rec_lines ls (line_num + 1) (line_string ^ (sprintf "  %s | %s\n" (padded_line_num line_num) l)) in
  let lines_before_string = rec_lines sr.lines_before (sr.line_num - (List.length sr.lines_before)) "" in
  let lines_after_string = rec_lines sr.lines_after (sr.line_num + 1) "" in
  String.concat [
    hdr;
    lines_before_string;
    sprintf "> %s | %s\n" (padded_line_num sr.line_num) sr.line;
    lines_after_string;
    ];;

let to_string (sr : t) = 
  match (sr.lines_before, sr.lines_after) with
  | ([],[]) -> single_line_to_string sr
  | _    -> multi_line_to_string sr;;
